internal/probing/parse_test.go

```diff
@@ -0,0 +1,94 @@
+package probing
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1alpha1 "package-operator.run/apis/core/v1alpha1"
+)
+
+func TestParse(t *testing.T) {
+	ctx := context.Background()
+	osp := []corev1alpha1.ObjectSetProbe{
+		{
+			Selector: corev1alpha1.ProbeSelector{
+				Kind: &corev1alpha1.PackageProbeKindSpec{
+					Kind:  "Test",
+					Group: "test",
+				},
+			},
+		},
+	}
+
+	p, err := Parse(ctx, osp)
+	require.NoError(t, err)
+	require.IsType(t, List{}, p)
+
+	if assert.Len(t, p, 1) {
```

Comment:
Wouldn't we ideally have a more specific test? For example:

```
	var probe Prober
	expected := List{&KindSelector{
		Prober: &StatusObservedGeneration{Prober: probe},
		GroupKind: schema.GroupKind{
			Group: osp[0].Selector.Kind.Group,
			Kind:  osp[0].Selector.Kind.Kind,
		},
	}}
	assert.True(t, reflect.DeepEqual(expected, p))
```

ps the example doesn't work and I don't know why :sweat_smile:

Reply:
Comparing interface slices with different pointers is not super trivial.
`reflect.DeepEqual` is respecting types, so a `*StatusObservedGeneration` != `Prober` even though one can be assigned to the other.

`assert.Equal` should handle this nicer (and gives you output for debugging!), but in my tests it tested for pointer equality within the list, raising a test error and printing:
```
        	            	expected: probing.list{(*probing.kindSelector)(0xc0001b2cc0)}
        	            	actual  : probing.list{(*probing.kindSelector)(0xc0001b2c60)}
```
Which is correct.
It really depends on the context, whether lists containing pointers to different objects are considered equal.

If in doubt, test and type-cast explicitly.

Reply:
Ohh wow, go is a very unforgiving language compared to python :D Thank you for the thorough explanation!!